src/libs/Middleware/HandleUnusedOptimisticID.ts

```diff
@@ -1,17 +1,36 @@
+import ONYXKEYS from '../../ONYXKEYS';
+import Report from '../../types/onyx/Report';
 import {Middleware} from '../Request';
 import * as PersistedRequests from '../actions/PersistedRequests';
 import deepReplaceKeysAndValues from '../deepReplaceKeysAndValues';
 
 const handleUnusedOptimisticID: Middleware = (requestResponse, request) =>
     requestResponse.then((response) => {
-        if (response && typeof response === 'object' && 'preexistingReportID' in response) {
+        const responseOnyxData = response?.onyxData ?? [];
+        responseOnyxData.forEach((onyxData) => {
+            const key = onyxData.key;
+            if (!key.startsWith(ONYXKEYS.COLLECTION.REPORT)) {
+                return;
+            }
+
+            if (!onyxData.value) {
+                return;
+            }
+
+            const report: Report = onyxData.value as Report;
+            const preexistingReportID = report.preexistingReportID;
+            if (!preexistingReportID) {
+                return;
+            }
             const oldReportID = request.data?.reportID;
-            PersistedRequests.getAll().forEach((persistedRequest, index) => {
-                // eslint-disable-next-line no-param-reassign
-                persistedRequest.data = deepReplaceKeysAndValues(persistedRequest.data, oldReportID as string, response.preexistingReportID as string);
-                PersistedRequests.update(index, persistedRequest);
-            });
-        }
+            PersistedRequests.getAll()
+                .slice(1)
```

Comment:
Do we need to slice by one request here? I assume this is done to ignore the current one but how can we guarantee that the current one is a persisted request?

Reply:
This is a great catch. We can use the third argument to the middleware function `isFromSequentialQueue` to determine this.

Reply:
Great catch. Updated